framework/src/meshgenerators/PlaneIDMeshGenerator.C

```diff
@@ -0,0 +1,108 @@
+
+  params.addRequiredParam<MeshGeneratorName>("input", "The mesh we want to modify");
+  params.addRequiredParam<std::vector<Real>>("plane_coordinates", "Coordinates of axial planes");
+  params.addParam<std::vector<unsigned int>>("num_ids_per_plane", "Number of unique ids per plane");
+  MooseEnum plane_axis("x y z", "z");
+  params.addParam<MooseEnum>("plane_axis", plane_axis, "Axis of plane");
+  params.addRequiredParam<std::string>("id_name", "Name of Integer ID set");
+  params.addParam<Real>("tolerance", 1.0E-4, "Tolerance for plane coordinate check");
+  params.addClassDescription(
+      "This PlaneIDMeshGenerator source code is to assigns plane extra ID for existing 3D meshes");
+
+  return params;
+}
+
+PlaneIDMeshGenerator::PlaneIDMeshGenerator(const InputParameters & parameters)
+  : MeshGenerator(parameters),
+    _input(getMesh("input")),
+    _axis_index(getParam<MooseEnum>("plane_axis")),
+    _element_id_name(getParam<std::string>("id_name"))
+{
+  if (!isParamValid("num_ids_per_plane"))
+    _planes = getParam<std::vector<Real>>("plane_coordinates");
+  else
+  {
+    _planes.clear();
+    std::vector<Real> base_planes = getParam<std::vector<Real>>("plane_coordinates");
+    std::vector<unsigned int> sublayers = getParam<std::vector<unsigned int>>("num_ids_per_plane");
+    if (base_planes.size() != sublayers.size() + 1)
+      paramError("plane_coordinates", "Sizes of 'z_planes' and '_sublayers' disagree");
+    _planes.push_back(base_planes[0]);
+    for (unsigned int i = 0; i < sublayers.size(); ++i)
+    {
+      Real layer_size = (base_planes[i + 1] - base_planes[i]) / (Real)sublayers[i];
+      for (unsigned int j = 0; j < sublayers[i]; ++j)
+        _planes.push_back(_planes.back() + layer_size);
+    }
+  }
+}
+
+std::unique_ptr<MeshBase>
+PlaneIDMeshGenerator::generate()
+{
+  std::unique_ptr<MeshBase> mesh = std::move(_input);
+  if (mesh->mesh_dimension() < _axis_index + 1)
+    paramError("plane_axis", "PlaneIDMeshGenerator must operate on a proper layer axis");
+  unsigned int eid = 0;
+  if (!mesh->has_elem_integer(_element_id_name))
+    eid = mesh->add_elem_integer(_element_id_name);
+  else
+    eid = mesh->get_elem_integer_index(_element_id_name);
+
+  Real tol = getParam<Real>("tolerance");
+  for (auto & elem : mesh->active_element_ptr_range())
+  {
+    int layer_id = getPlaneID(elem->centroid());
+    if (layer_id == -1 || layer_id == (int)_planes.size())
+      mooseError("The axial layers do not cover element at ", elem->centroid());
+    for (unsigned int i = 0; i < elem->n_nodes(); ++i)
+    {
+      const Point & p = elem->point(i) - (elem->point(i) - elem->centroid()) * tol;
+      if (getPlaneID(p) != layer_id)
+        mooseError("Element at ", elem->centroid(), " is cut by the axial layers");
+    }
+    elem->set_extra_integer(eid, layer_id);
+  }
+  return dynamic_pointer_cast<MeshBase>(mesh);
+}
+
+int
+PlaneIDMeshGenerator::getPlaneID(const Point & p) const
+{
+  if (_planes.size() == 0)
```

Comment:
Why would anyone ever use this with no planes provided? If planes aren't provided, I think it should error in the constructor